src/stash/merge.rs

```diff
@@ -0,0 +1,408 @@
+// LNP/BP Rust Library
+// Written in 2020 by
+//     Rajarshi Maitra <rajarshi149@protonmail.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the MIT License
+// along with this software.
+// If not, see <https://opensource.org/licenses/MIT>.
+
+use crate::contract::nodes::{Extension, Genesis, Node, Transition};
+use crate::contract::{Assignments, OwnedState, StateTypes};
+use lnpbp::client_side_validation::{
+    CommitConceal, CommitEncode, ConsensusCommit,
+};
+/// A trait to merge two structures modifying the revealed status
+/// of the first one.
+///
+/// The resulting structure will depend on the openness of the both the input.
+/// And the most revealed condition among the two will be selected
+/// Usage: prevent hiding already known previous state data by merging
+/// incoming new consignment in stash.
+///
+/// The follwoing conversion logic is intended by this trait:
+///
+/// merge (Revelaed, Anything) = Revealed
+/// merge(ConfidentialSeal, ConfidentiualAmount) = Revealed
+/// merge(ConfidentialAmount, ConfidentialSeal) = Revealed
+/// merge(Confidential, Anything) = Anything
+///   
+pub trait MergeRevealed: Sized {
+    fn merge_revealed(&mut self, other: &Self) -> Result<bool, String>;
+}
+
+impl<STATE> MergeRevealed for OwnedState<STATE>
+where
+    Self: Clone,
+    STATE: StateTypes,
+    STATE::Confidential: PartialEq + Eq,
+    STATE::Confidential:
+        From<<STATE::Revealed as CommitConceal>::ConcealedCommitment>,
+{
+    fn merge_revealed(&mut self, other: &Self) -> Result<bool, String> {
+        // if self and other is different through error
+        if self.commit_serialize() != other.commit_serialize() {
+            Err(s!("Owned state data missmatch in merge operation"))
+        } else {
+            match (self.clone(), other) {
+                // Anything + Revealed = Revealed
+                (
+                    _,
+                    OwnedState::Revealed {
+                        seal_definition: seal,
+                        assigned_state: state,
+                    },
+                ) => {
+                    *self = OwnedState::Revealed {
+                        seal_definition: seal.clone(),
```

Comment:
Hm.. this will not work without clone... makes the whole idea of mutating self to avoid cloning quite useless...

Seems we need to consume `self` and return it back, making function signature to be `into_revealed(mut self, other: Self) -> Result<Self, Self::Error>` (and the trait can be renamed as `IntoRevealed`)

Reply:
Updating trait signature as suggested.